aioxmpp/cache.py

```diff
@@ -52,25 +103,26 @@ class LRUDict(collections.abc.MutableMapping):
 
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
-        self.__data = {}
-        self.__data_used = {}
+        self.__links = {}
+        self.__root = _init_linked_list()
+
         self.__maxsize = 1
-        self.__ctr = 0
-
-    def _purge_old(self, n):
-        keys_in_age_order = sorted(
-            self.__data_used.items(),
-            key=lambda x: x[1]
-        )
-        keys_to_delete = keys_in_age_order[:n]
-        for key, _ in keys_to_delete:
-            del self.__data[key]
-            del self.__data_used[key]
-        keys_to_keep = keys_in_age_order[n:]
-        # avoid the counter becoming large
-        self.__ctr = len(keys_to_keep)
-        for i, (key, _) in enumerate(keys_to_keep):
-            self.__data_used[key] = i
+
+    def _test_consistency(self):
+        """
+        This method is only used for testing to assert that the operations
+        leave the LRUDict in a valid state.
+        """
+        return (_length(self.__root) == len(self.__links) and
+                _has_consistent_links(self.__root, self.__links))
+
+    def _purge(self):
+        if self.__maxsize is None:
+            return
+
+        while len(self.__links) > self.__maxsize:
+            link = _remove_link(self.__root.prev)
+            del self.__links[link.key]
```

Comment:
This is clearly more optimized for the common case that an LRUDict is full. :+1: